File: App_Clint/lib/features/booking/domain/entities/booking.dart
import 'package:equatable/equatable.dart';

/// Booking entity representing a user's booking in the domain layer
class Booking extends Equatable {
  final int id;
  final int userId;
  final String userName;
  final int companyId;
  final String companyName;
  final int workerId;
  final String workerName;
  final int workerWorkTypeId;
  final String? workTypeName;
  final DateTime bookingDate;
  final DateTime? startDate;
  final DateTime? endDate;
  final String? startDateDisplay;
  final String? endDateDisplay;
  final String address;
  final int? userLocationId;
  final String? locationName;
  final double? lat;
  final double? lng;
  /// Numeric status codes: see `BookingStatusCodes`.
  final int status;
  final DateTime createdAt;
  final String? rejectionReason;
  final double servicePrice;
  final double platformFeeAmount;
  final double totalPrice;
  final bool isMonthlyPricing;
  final String? paymentMethod;
  final bool isWorkerArrivalConfirmed;
  final DateTime? workerArrivalConfirmedAt;
  final bool walletAmountReserved;
  final bool walletAmountCaptured;
  final DateTime? walletCapturedAt;

  const Booking({
    required this.id,
    required this.userId,
    required this.userName,
    required this.companyId,
    required this.companyName,
    required this.workerId,
    required this.workerName,
    required this.workerWorkTypeId,
    this.workTypeName,
    required this.bookingDate,
    this.startDate,
    this.endDate,
    this.startDateDisplay,
    this.endDateDisplay,
    required this.address,
    this.userLocationId,
    this.locationName,
    this.lat,
    this.lng,
    required this.status,
    required this.createdAt,
    this.rejectionReason,
    this.servicePrice = 0,
    this.platformFeeAmount = 0,
    this.totalPrice = 0,
    this.isMonthlyPricing = false,
    this.paymentMethod,
    this.isWorkerArrivalConfirmed = false,
    this.workerArrivalConfirmedAt,
    this.walletAmountReserved = false,
    this.walletAmountCaptured = false,
    this.walletCapturedAt,
  });

  bool get hasStoredPricing => totalPrice > 0 || servicePrice > 0;

  bool get isWalletPayment =>
      paymentMethod?.trim().toLowerCase() == 'wallet' ||
      walletAmountReserved ||
      walletAmountCaptured;

  /// Normalized key for UI / filters (matches l10n keys where applicable).
  String get statusString {
    switch (status) {
      case 0:
        return 'pending';
      case 1:
        return 'approved';
      case 2:
        return 'on_the_way';
      case 3:
        return 'completed';
      case 4:
        return 'canceled';
      case 5:
        return 'rejected';
      default:
        return 'pending';
    }
  }

  Booking copyWith({
    int? status,
    String? rejectionReason,
    bool? isWorkerArrivalConfirmed,
    DateTime? workerArrivalConfirmedAt,
    bool? walletAmountReserved,
    bool? walletAmountCaptured,
    DateTime? walletCapturedAt,
  }) {
    return Booking(
      id: id,
      userId: userId,
      userName: userName,
      companyId: companyId,
      companyName: companyName,
      workerId: workerId,
      workerName: workerName,
      workerWorkTypeId: workerWorkTypeId,
      workTypeName: workTypeName,
      bookingDate: bookingDate,
      startDate: startDate,
      endDate: endDate,
      startDateDisplay: startDateDisplay,
      endDateDisplay: endDateDisplay,
      address: address,
      userLocationId: userLocationId,
      locationName: locationName,
      lat: lat,
      lng: lng,
      status: status ?? this.status,
      createdAt: createdAt,
      rejectionReason: rejectionReason ?? this.rejectionReason,
      servicePrice: servicePrice,
      platformFeeAmount: platformFeeAmount,
      totalPrice: totalPrice,
      isMonthlyPricing: isMonthlyPricing,
      paymentMethod: paymentMethod,
      isWorkerArrivalConfirmed:
          isWorkerArrivalConfirmed ?? this.isWorkerArrivalConfirmed,
      workerArrivalConfirmedAt:
          workerArrivalConfirmedAt ?? this.workerArrivalConfirmedAt,
      walletAmountReserved: walletAmountReserved ?? this.walletAmountReserved,
      walletAmountCaptured: walletAmountCaptured ?? this.walletAmountCaptured,
      walletCapturedAt: walletCapturedAt ?? this.walletCapturedAt,
    );
  }

  @override
  List<Object?> get props => [
        id,
        userId,
        userName,
        companyId,
        companyName,
        workerId,
        workerName,
        workerWorkTypeId,
        workTypeName,
        bookingDate,
        startDate,
        endDate,
        startDateDisplay,
        endDateDisplay,
        address,
        userLocationId,
        locationName,
        lat,
        lng,
        status,
        createdAt,
        rejectionReason,
        servicePrice,
        platformFeeAmount,
        totalPrice,
        isMonthlyPricing,
        paymentMethod,
        isWorkerArrivalConfirmed,
        workerArrivalConfirmedAt,
        walletAmountReserved,
        walletAmountCaptured,
        walletCapturedAt,
      ];
}

File: Backend API/CleaningHouse_API/Models/Customers/Booking.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using CleaningHouse_API.Models.Common;
using CleaningHouse_API.Models.Companies;

namespace CleaningHouse_API.Models.Customers;

[Table("Bookings")]
public class Booking
{
    [Key]
    public int Id { get; set; }

    [Required]
    public int UserId { get; set; }

    [Required]
    public int CompanyId { get; set; }

    [Required]
    public int WorkerId { get; set; }
    
    [Required]
    public int WorkTypeId { get; set; }

    [Required]
    public DateTime BookingDate { get; set; }

    [Required]
    [MaxLength(50)]
    public string StartDate { get; set; } = string.Empty;

    [Required]
    [MaxLength(50)]
    public string EndDate { get; set; } = string.Empty;

    [Required]
    [MaxLength(500)]
    public string Address { get; set; } = string.Empty;

    public int? UserLocationId { get; set; }

    [Required]
    public int Status { get; set; }

    [MaxLength(2000)]
    public string? RejectionReason { get; set; }

    [Required]
    [Column(TypeName = "decimal(18,2)")]
    public decimal ServicePrice { get; set; }

    [Required]
    [Column(TypeName = "decimal(18,2)")]
    public decimal PlatformFeeAmount { get; set; }

    [Required]
    [Column(TypeName = "decimal(18,2)")]
    public decimal TotalPrice { get; set; }

    public bool IsMonthlyPricing { get; set; }

    public bool IsWorkerArrivalConfirmed { get; set; }

    public DateTime? WorkerArrivalConfirmedAt { get; set; }

    public bool WalletAmountReserved { get; set; }

    public bool WalletAmountCaptured { get; set; }

    public DateTime? WalletCapturedAt { get; set; }

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    // Navigation Properties
    [ForeignKey("UserId")]
    public virtual AppUser? AppUser { get; set; }

    [ForeignKey("CompanyId")]
    public virtual Company? Company { get; set; }

    [ForeignKey("WorkerId")]
    public virtual Worker? Worker { get; set; }
    
    [ForeignKey("WorkTypeId")]
    public virtual WorkType? WorkType { get; set; }

    [ForeignKey(nameof(UserLocationId))]
    public virtual UserLocation? UserLocation { get; set; }

    public virtual ICollection<Payment>? Payments { get; set; }
    public virtual ICollection<Review>? Reviews { get; set; }
}



